### DISCUSSION TITLE
MODE CTS Output Format

### DISCUSSION THREAD

**Discussion by ebower084** [dtcenter/METplus#1681] [Statistical Computation] [answered]
Good afternoon,

My colleagues and I are using MODE for some object based verification, and we came across a question about the output format that we weren't quite able to figure out from the User Guide.

In the cts.txt output file from MODE, there are the two rows, "raw" and "object." Are the statistics for the "object" line computed using only the matched pairs of objects from the forecast and observed fields? Or are these metrics computed using all objects in the fields whether they are matched or not?

Thank you for your help!
EB

**Comment by JohnHalleyGotway** [comment, marked as answer]
Hello,

I see you have a question about interpreting the Contingency Table Statistics (CTS) output generated by the MODE tool. As you note, the *_cts.txt output file contains 2 lines of data, one where the `FIELD` column is set to `RAW` and another for `OBJECT`. These are referenced in [this section](https://met.readthedocs.io/en/latest/Users_Guide/mode.html#mode-output) of the MET User's Guide.

Let me clarify that these contingency table counts and statistics are computed grid point by grid point. They DO NOT take into account MODE's object-based matching and merging logic. Instead, they represent the traditional way of doing verification, grid point by grid point.

So why does MODE write this information if it has nothing to do with object-based verification? When MODE was first released, the idea of object-based verification was very new. It represented a large departure from traditional scoring methods. That motivated us to include those traditional categorical scores in MODE's output to make life easier for users. They could compare the utility of traditional statistics to the (hopefully) more useful object comparisons in the *_obj.txt files from MODE.

Big picture, to compute a grid-to-grid analysis over a flexible set of masking regions and leveraging neighborhood methods, I'd recommend running Grid-Stat rather than using the CTS output from MODE.

To be clear, the `RAW` output is generated by:
1. Reading the raw fcst/obs fields.
2. Applying the `fcst.conv_thresh` threshold to the forecast field and `obs.conv_thresh` threshold to the observation field to define 0's and 1.s
3. Scoring those 0's and 1's grid point by grid point to construct a 2x2 contingency table.
4. Deriving `RAW` CTS statistics from that contingency table.

The scores for the `OJBECT` output are very similar. However instead of scoring the thresholded raw values, it scores the resolved object fields. Any grid point included in an object is counted as an event. Any grid point not included in an object is a non-event. But this is still scored grid point by grid point, not object by object.

Note however, that the METviewer tool does include support for deriving object-based contingency table statistics, like these:
<img width="115" alt="Screen Shot 2022-06-29 at 12 46 47 PM" src="https://user-images.githubusercontent.com/21087144/176512810-71fc559f-49ae-454a-8436-68f352a3d06b.png">

In addition, the existing [mode_summary.R](https://github.com/dtcenter/MET/blob/main_v10.1/met/scripts/Rscripts/mode_summary.R) script can provide a nice summary of multiple MODE runs. An example of its output can be found at the bottom of this [webpage](https://dtcenter.org/community-code/model-evaluation-tools-met/sample-analysis-scripts).

Hope that helps.

**Reply by ebower084** [reply]
Thank you John, this was very helpful! One last question for clarification if you don't mind? Any grid point in any object, matched or unmatched, is included in the calculation of the "OBJECT" line statistics?
Thanks,
EB

**Reply by JohnHalleyGotway** [reply]
Yes, that's correct. Any grid point included in an object is counted as an event in contingency table regardless of that object's matching status. To illustrate, I ran the ncview utility to plot the fcst_obj_id and obs_obj_id variables from a sample NetCDF output file from MODE. These are the fields that are scored using the object contingency table and reported in the OBJECT line of the cts.txt output file.

<img width="751" alt="Screen Shot 2022-07-05 at 9 09 06 AM" src="https://user-images.githubusercontent.com/21087144/177359750-eca870c3-b796-4d3c-aa29-2b76312572e3.png">

**Reply by ebower084** [reply]
Great, thank you for all of your help!
EB

**Reply by JohnHalleyGotway** [reply]
I'll go ahead and mark this discussion as being answered and then lock conversation on it.
That prevents us from losing track of future posts. If more questions arise in the future, please feel free to create a new discussion.
Thanks
